test_noobaa_space_available_using_cli fails in 4.21 - investigate and fix

Relevant runs : 

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/994/46191/2079991/2080156/log

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/994/46216/2081535/2081699/log

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/994/46286/2083761/2083969/log

Backtrace : 
self = <mcg.test_noobaa_metrics.TestNoobaaMetrics object at 0x7f69830b8e10>
mcg_obj = <ocs_ci.ocs.resources.mcg.MCG object at 0x7f696aad90d0>
awscli_pod_session = <ocs_ci.ocs.resources.pod.Pod object at 0x7f698fcf3950>
bucket_factory = <function bucket_factory_fixture.._create_buckets at 0x7f68e1730040>
test_directory_setup = SetupDirs(origin_dir='test_noobaa_space_available_using_cli/origin', result_dir='test_noobaa_space_available_using_cli/result')


@config.run_with_provider_context_if_available
def test_noobaa_space_available_using_cli(
    self, mcg_obj, awscli_pod_session, bucket_factory, test_directory_setup
):
    """
    Test that 'bucket update' command with --max-size and --max-object parameters work as expected
    1. Create bucket and check the initial quota status
    2. Set the max quota size and verify that it works as expected
    3. Update the max-objects parameter and then
        a. Write number of objects almost reaching max-objects, verify correct quota status
        b. Write number of objects exactly equal to max-objects, verify correct quota status
        c. Try to write one object more, verify that the command fails as expected
    4. Update the max-objects parameter to a bigger number, write another object and verify that this succeeds
    Args:
        mcg_obj (obj): An object representing the current state of the MCG in the cluster
        awscli_pod_session (pod): A pod running the AWSCLI tools
        bucket_factory: Calling this fixture creates a new bucket(s)
    """

    # 1. Create bucket and check original quota status

  bucket_name = bucket_factory(amount=1, interface="CLI")[0].name

/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/tests/functional/object/mcg/test_noobaa_metrics.py:57: 



/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/tests/conftest.py:3571: in create_buckets
    created_bucket = BUCKET_MAP[interface.lower()](
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/ocs/resources/objectbucket.py:391: in _init
    self.mcg.exec_mcg_cmd(
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/ocs/resources/mcg.py:945: in exec_mcg_cmd
    result = self.odf_cli_runner.run_noobaa(
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/helpers/odf_cli.py:345: in run_noobaa
    output = exec_cmd(full_command, ignore_error=ignore_error, *kwargs)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/utility/retry.py:79: in f_retry
    return f(args, **kwargs)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/utility/utils.py:837: in exec_cmd
    completed_process = subprocess.run(
/usr/lib64/python3.11/subprocess.py:550: in run
    stdout, stderr = process.communicate(input, timeout=timeout)
/usr/lib64/python3.11/subprocess.py:1209: in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
/usr/lib64/python3.11/subprocess.py:2116: in _communicate
    self._check_timeout(endtime, orig_timeout, stdout, stderr)



self = <Popen: returncode: -9 args: ['odf', 'noobaa', 'obc', 'create', '--exact', '...>
endtime = 773160.499789621, orig_timeout = 600, stdout_seq = []
stderr_seq = [b'time="2026-03-24T20:59:06-04:00" level=info msg="\xe2\x9c\x85 Exists: StorageClass \"openshift-storage.noobaa.io\...04:00" level=info msg="\xe2\x8f\xb3 OBC \"cli-bucket-c3ad9719d2514dd79a3943f1c2e32\" Phase is \"Pending\""\n', ...]
skip_check_and_raise = False


def _check_timeout(self, endtime, orig_timeout, stdout_seq, stderr_seq,
                   skip_check_and_raise=False):
    """Convenience for checking if a timeout has expired."""
    if endtime is None:
        return
    if skip_check_and_raise or _time() > endtime:

      raise TimeoutExpired(
                self.args, orig_timeout,
                output=b''.join(stdout_seq) if stdout_seq else None,
                stderr=b''.join(stderr_seq) if stderr_seq else None)
E           subprocess.TimeoutExpired: Command '['odf', 'noobaa', 'obc', 'create', '--exact', 'cli-bucket-c3ad9719d2514dd79a3943f1c2e32', '-n', 'openshift-storage']' timed out after 600 seconds



/usr/lib64/python3.11/subprocess.py:1253: TimeoutExpired

--- a/tests/functional/object/mcg/test_noobaa_metrics.py
+++ b/tests/functional/object/mcg/test_noobaa_metrics.py
@@ -1,5 +1,5 @@
 import logging
-import time
+
 from ocs_ci.framework import config
 from ocs_ci.framework.pytest_customization.marks import (
     tier2,
@@ -9,6 +9,7 @@ from ocs_ci.framework.pytest_customization.marks import (
 )
 
 from ocs_ci.ocs.exceptions import CommandFailed
+from ocs_ci.utility.utils import TimeoutSampler
 
 from ocs_ci.ocs.bucket_utils import (
     get_bucket_status_value,
@@ -16,6 +17,31 @@ from ocs_ci.ocs.bucket_utils import (
 )
 
 logger = logging.getLogger(__name__)
+
+
+def wait_for_num_objects(mcg_obj, bucket_name, expected_count, timeout=360):
+    """
+    Poll bucket status until the number of objects reaches the expected count.
+
+    Args:
+        mcg_obj (obj): An object representing the current state of the MCG in the cluster
+        bucket_name (str): Name of the bucket to check
+        expected_count (int): Expected number of objects
+        timeout (int): Maximum time to wait in seconds
+
+    """
+    logger.info(f"Waiting for object count to reach {expected_count}")
+    for num_objects in TimeoutSampler(
+        timeout=timeout,
+        sleep=30,
+        func=get_bucket_status_value,
+        mcg_obj=mcg_obj,
+        bucket_name=bucket_name,
+        key="Num Objects",
+    ):
+        if int(num_objects) == expected_count:
+            logger.info(f"Object count reached {num_objects}")
+            break
 
 
 class QuotaStatus:
@@ -51,6 +77,7 @@ class TestNoobaaMetrics:
             mcg_obj (obj): An object representing the current state of the MCG in the cluster
             awscli_pod_session (pod): A pod running the AWSCLI tools
             bucket_factory: Calling this fixture creates a new bucket(s)
+            test_directory_setup: Fixture that sets up origin and result directories for the test
         """
 
         # 1. Create bucket and check original quota status
@@ -92,15 +119,22 @@ class TestNoobaaMetrics:
             pattern="ObjKey1-",
             mcg_obj=mcg_obj,
         )
-        time.sleep(240)
 
-        space_avail_after_write = get_bucket_status_value(
-            mcg_obj, bucket_name, "Data Space Avail"
-        )
-        space_avail_after_write = float(space_avail_after_write.split(" ")[0])
-        assert (
-            space_avail_after_write < space_avail_after_update
-        ), f"Available space before write = {space_avail_after_update}, after write = {space_avail_after_write}"
+        logger.info("Waiting for space available to decrease after writing 1 object")
+        for space_avail_after_write in TimeoutSampler(
+            timeout=360,
+            sleep=30,
+            func=get_bucket_status_value,
+            mcg_obj=mcg_obj,
+            bucket_name=bucket_name,
+            key="Data Space Avail",
+        ):
+            space_avail_after_write = float(space_avail_after_write.split(" ")[0])
+            if space_avail_after_write < space_avail_after_update:
+                logger.info(
+                    f"Space available decreased: {space_avail_after_update} -> {space_avail_after_write}"
+                )
+                break
 
         # 3. Update bucket with --max-objects quota and verify that it worked as expected
         max_objects = 10
@@ -127,16 +161,8 @@ class TestNoobaaMetrics:
             mcg_obj=mcg_obj,
         )
 
-        time.sleep(180)
+        wait_for_num_objects(mcg_obj, bucket_name, max_objects - 1)
 
-        # Verify that max_objects -1 have been copied,
-        # only 1 object is available and quota status is QuotaStatus.Approaching
-        num_objects_after_write = get_bucket_status_value(
-            mcg_obj, bucket_name, "Num Objects"
-        )
-        assert (
-            int(num_objects_after_write) == max_objects - 1
-        ), f"Number of written objects is {num_objects_after_write}, expected {max_objects -1}"
         num_objects_avail_after_write = get_bucket_status_value(
             mcg_obj, bucket_name, "Num Objects Avail"
         )
@@ -159,14 +185,9 @@ class TestNoobaaMetrics:
             pattern="ObjKey3-",
             mcg_obj=mcg_obj,
         )
-        time.sleep(180)
 
-        num_objects_after_write = get_bucket_status_value(
-            mcg_obj, bucket_name, "Num Objects"
-        )
-        assert (
-            int(num_objects_after_write) == max_objects
-        ), f"Number of written objects is {num_objects_after_write}, expected {max_objects}"
+        wait_for_num_objects(mcg_obj, bucket_name, max_objects)
+
         num_objects_avail_after_write = get_bucket_status_value(
             mcg_obj, bucket_name, "Num Objects Avail"
         )
@@ -213,8 +234,18 @@ class TestNoobaaMetrics:
             pattern="ObjKey5-",
             mcg_obj=mcg_obj,
         )
-        time.sleep(180)
-        quota_status = get_bucket_status_value(mcg_obj, bucket_name, "QuotaStatus")
-        assert (
-            quota_status == QuotaStatus.OPTIMAL
-        ), f"Quota status after update {quota_status}, expected {QuotaStatus.OPTIMAL}"
+
+        logger.info(
+            "Waiting for quota status to return to OPTIMAL after increasing max-objects"
+        )
+        for quota_status in TimeoutSampler(
+            timeout=360,
+            sleep=30,
+            func=get_bucket_status_value,
+            mcg_obj=mcg_obj,
+            bucket_name=bucket_name,
+            key="QuotaStatus",
+        ):
+            if quota_status == QuotaStatus.OPTIMAL:
+                logger.info("Quota status returned to OPTIMAL")
+                break

--- a/tests/functional/object/mcg/test_noobaa_metrics.py
+++ b/tests/functional/object/mcg/test_noobaa_metrics.py
@@ -8,7 +8,7 @@ from ocs_ci.framework.pytest_customization.marks import (
     mcg,
 )
 
-from ocs_ci.ocs.exceptions import CommandFailed
+from ocs_ci.ocs.exceptions import CommandFailed, TimeoutExpiredError
 from ocs_ci.utility.utils import TimeoutSampler
 
 from ocs_ci.ocs.bucket_utils import (
@@ -31,17 +31,25 @@ def wait_for_num_objects(mcg_obj, bucket_name, expected_count, timeout=360):
 
     """
     logger.info(f"Waiting for object count to reach {expected_count}")
-    for num_objects in TimeoutSampler(
-        timeout=timeout,
-        sleep=30,
-        func=get_bucket_status_value,
-        mcg_obj=mcg_obj,
-        bucket_name=bucket_name,
-        key="Num Objects",
-    ):
-        if int(num_objects) == expected_count:
-            logger.info(f"Object count reached {num_objects}")
-            break
+    last_seen = None
+    try:
+        for num_objects in TimeoutSampler(
+            timeout=timeout,
+            sleep=30,
+            func=get_bucket_status_value,
+            mcg_obj=mcg_obj,
+            bucket_name=bucket_name,
+            key="Num Objects",
+        ):
+            last_seen = num_objects
+            if int(num_objects) == expected_count:
+                logger.info(f"Object count reached {num_objects}")
+                break
+    except TimeoutExpiredError:
+        assert False, (
+            f"Object count did not reach {expected_count} within {timeout}s, "
+            f"last observed: {last_seen}"
+        )
 
 
 class QuotaStatus:
@@ -121,20 +129,28 @@ class TestNoobaaMetrics:
         )
 
         logger.info("Waiting for space available to decrease after writing 1 object")
-        for space_avail_after_write in TimeoutSampler(
-            timeout=360,
-            sleep=30,
-            func=get_bucket_status_value,
-            mcg_obj=mcg_obj,
-            bucket_name=bucket_name,
-            key="Data Space Avail",
-        ):
-            space_avail_after_write = float(space_avail_after_write.split(" ")[0])
-            if space_avail_after_write < space_avail_after_update:
-                logger.info(
-                    f"Space available decreased: {space_avail_after_update} -> {space_avail_after_write}"
-                )
-                break
+        last_seen = None
+        try:
+            for space_avail_after_write in TimeoutSampler(
+                timeout=360,
+                sleep=30,
+                func=get_bucket_status_value,
+                mcg_obj=mcg_obj,
+                bucket_name=bucket_name,
+                key="Data Space Avail",
+            ):
+                space_avail_after_write = float(space_avail_after_write.split(" ")[0])
+                last_seen = space_avail_after_write
+                if space_avail_after_write < space_avail_after_update:
+                    logger.info(
+                        f"Space available decreased: {space_avail_after_update} -> {space_avail_after_write}"
+                    )
+                    break
+        except TimeoutExpiredError:
+            assert False, (
+                f"Space available did not decrease below {space_avail_after_update} "
+                f"within 360s, last observed: {last_seen}"
+            )
 
         # 3. Update bucket with --max-objects quota and verify that it worked as expected
         max_objects = 10
@@ -238,14 +254,22 @@ class TestNoobaaMetrics:
         logger.info(
             "Waiting for quota status to return to OPTIMAL after increasing max-objects"
         )
-        for quota_status in TimeoutSampler(
-            timeout=360,
-            sleep=30,
-            func=get_bucket_status_value,
-            mcg_obj=mcg_obj,
-            bucket_name=bucket_name,
-            key="QuotaStatus",
-        ):
-            if quota_status == QuotaStatus.OPTIMAL:
-                logger.info("Quota status returned to OPTIMAL")
-                break
+        last_seen = None
+        try:
+            for quota_status in TimeoutSampler(
+                timeout=360,
+                sleep=30,
+                func=get_bucket_status_value,
+                mcg_obj=mcg_obj,
+                bucket_name=bucket_name,
+                key="QuotaStatus",
+            ):
+                last_seen = quota_status
+                if quota_status == QuotaStatus.OPTIMAL:
+                    logger.info("Quota status returned to OPTIMAL")
+                    break
+        except TimeoutExpiredError:
+            assert False, (
+                f"Quota status did not reach {QuotaStatus.OPTIMAL} "
+                f"within 360s, last observed: {last_seen}"
+            )

